pkg/infra/BlockCollector_test.go

```diff
@@ -0,0 +1,56 @@
+package infra_test
+
+import (
+	"tape/pkg/infra"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("BlockCollector", func() {
+
+	Context("NewBlockCollector", func() {
+		It("Should create New BlockCollector", func() {
+			instance := infra.NewBlockCollector(1)
+			Expect(instance).NotTo(BeNil())
+			Expect(instance.Observers).Should(Equal(1))
+			Expect(instance.Registry).NotTo(BeNil())
+		})
+	})
+
+	Context("Commit", func() {
+		It("should return true when observer is 1", func() {
+			instance := infra.NewBlockCollector(1)
+			Expect(instance.Commit(uint64(1))).To(BeTrue())
+		})
+
+		It("Should added when the block committed at 1st time", func() {
+			instance := infra.NewBlockCollector(2)
+			Expect(instance.Commit(uint64(1))).To(BeFalse())
+			i, ok := instance.Registry[1]
+			Expect(ok).To(BeTrue())
+			Expect(i).To(Equal(1))
+		})
+
+		It("Should removed when the block committed at n time", func() {
+			instance := infra.NewBlockCollector(2)
+			Expect(instance.Commit(uint64(1))).To(BeFalse())
+			Expect(instance.Commit(uint64(1))).To(BeTrue())
+			Expect(len(instance.Registry)).Should(Equal(0))
+		})
+
+		It("Should supports multiple running", func() {
+			instance := infra.NewBlockCollector(3)
+			go instance.Commit(uint64(1))
+			go instance.Commit(uint64(1))
+			time.Sleep(time.Millisecond * 100)
+			Expect(len(instance.Registry)).Should(Equal(1))
+			i, ok := instance.Registry[1]
+			Expect(ok).To(BeTrue())
+			Expect(i).To(Equal(1))
+			instance.Commit(uint64(1))
+			Expect(len(instance.Registry)).Should(Equal(0))
+		})
+	})
+})
```

Comment:
```go
package infra_test

import (
	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"
	"sync"
	"tape/pkg/infra"
)

var _ = Describe("BlockCollector", func() {

	Context("Commit", func() {
		It("should return true when observer is 1", func() {
			instance := infra.NewBlockCollector(1)
			Expect(instance.Commit(uint64(1))).To(BeTrue())
		})

		It("Should added when the block committed at 1st time", func() {
			instance := infra.NewBlockCollector(3)
			Expect(instance.Commit(uint64(1))).To(BeFalse())
			Expect(instance.Commit(uint64(1))).To(BeFalse())
			Expect(instance.Commit(uint64(1))).To(BeTrue())
		})

		It("Should supports multiple running", func() {
			instance := infra.NewBlockCollector(100)
			signal := make(chan struct{})
			var wg sync.WaitGroup
			for i := 0; i < 100; i++ {
				wg.Add(1)
				go func() {
					defer wg.Done()

					if instance.Commit(uint64(1)) {
						close(signal)
					}
				}()
			}

			Expect(signal).To(BeClosed())
		})
	})
})
```
Glad you covered parallel commit case. Some suggestions:
- test api, not implementation details: test should not be modified during refactor
- avoid sleep in test